Automatically mute users who attempt to post a slur, fixes #8062

abatula opened the pull request:
Fixes #8062
### Changes

Work in progress. 
- [x]  Block user from posting a message containing  a word contained in an array of blacklisted words
- [x] Permanently revoke chat privileges
- [x] Faster/more elegant method for checking all banned words
- [x] Load in the official list of banned slurs
- [x] Is there only one file, or are there ones for different languages?
- [x] Email the moderators
- [x] Send slack message to moderators

Tests:
- [x] Blocks posting chat with a slur
- [x] Blocks posting in private group
- [x] Revokes chat privileges
- [x] Sends email to moderators
- [x] Sends slack notification to moderators

---

UUID: f5a68145-b375-4b99-8fb4-f9ff7a2c52b5

abatula committed:
First attempt at the muteUser function: revokes user's chat privileges and notifies moderators

abatula committed:
Manual merge for cherry-pick

abatula committed:
Initial psuedo-code for checking for slurs in messages

abatula committed:
Initial working prototype for blocking posting of slurs. Moved check from group.js to the chat api. Still needs: to permanently revoke chat privileges, to notify the moderators, a better method for checking for the blacklisted words, and a way to get the real list of words to check.

abatula committed:
Initial working prototype for blocking posting of slurs. Moved check from group.js to the chat api. Still needs: to permanently revoke chat privileges, to notify the moderators, a better method for checking for the blacklisted words, and a way to get the real list of words to check.

abatula committed:
Permanently revoke chat privileges when attempting to post a slur.

abatula committed:
Permanently revoke chat privileges when attempting to post a slur.

abatula committed:
Removed console logs

abatula committed:
Removed console logs

abatula committed:
Created report to be sent to moderators via email

abatula commented:
Progress update:
- Does a simple check for banned words in a list
- Revokes chat privileges and prevents posting if a match is found in the message
- Created report to send to the moderators (written but email is untested)

TheHollidayInn commented on website/server/controllers/api-v3/chat.js:
```diff
@@ -100,6 +100,32 @@ api.postChat = {
+      if (tmp > -1) {
+        // Need a way to make revoking chat permanent
+        user.muteUser(message, groupId);
+        throw new NotFound('Your message contained inapropriate language, and your chat privileges have been revoked.');
```
This should be moved to a local. Use `res.t()`.

Also, you should move this check below, because you need to check the privacy setting of the group. Private groups can use banned language.

abatula committed:
Do not moderate private groups

abatula committed:
Do not moderate private groups

abatula committed:
Moved slur check to a generic check for banned words function

abatula committed:
Moved slur check to a generic check for banned words function

Alys commented:
@abatula I'm sorry but can you change this back so that it **does** apply to private guilds and parties?

It's true that people can swear in a private group if the group allows it, however slurs that violate the Terms of Service are prohibited everywhere.

abatula committed:
Moved list of slurs to a separate file, fixed misplacement of return in ContainsBannedWords() function

abatula committed:
Moved list of slurs to a separate file, fixed misplacement of return in ContainsBannedWords() function

abatula committed:
Slurs are blocked in both public and private groups

abatula committed:
Slurs are blocked in both public and private groups

abatula committed:
Added code to send a slack message for slurs

abatula committed:
Added code to send a slack message for slurs

abatula committed:
Fixed formatting issues

abatula committed:
Fixed formatting issues

abatula commented:
Progress update:
- Restored blocking of slurs in private groups
- Created generic function to check for banned words
  - Needs to coordinate with #8197 
  - Is the chat.js file an appropriate place for this function? 
- List of banned slurs is loaded from a file in `website/common` 
- Initial code to send a slack message

abatula committed:
Incorporated tectContainsBannedWords() function from PR 8197, added an argument to specify the list of banned words to check

abatula committed:
Incorporated tectContainsBannedWords() function from PR 8197, added an argument to specify the list of banned words to check

abatula commented:
I incorporated the textContainsBannedWords() function from #8197, and added an additional input to specify the list of banned words to test. 

I'm not sure of the best way to share one specific change between two in-progress pull requests without cluttering one of them up with all of the other's changes. A third pull request that these two both pull in/push to? Make sure we're using the same function, and when it's finished delete one of them or hope git is smart enough to realize we both added the exact same code?

abatula commented:
@TheHollidayInn Are you planning to push your changes to this pull request? Or should I pull in your branch? Our two pull requests are similar but not totally the same, and I want to make sure you get the credit for your code.

TheHollidayInn commented:
I believe you moved the code over, correct?

If you want, it looks like you could still pull over my integration tests. Don't worry about crediting.

abatula commented:
Oh, no! Sorry for the misunderstanding. I only copied over your `textContainsBannedWords()` function, because both of our code sections are supposed to use it. I intended to delete that from my code once yours was finished and merged. My code only looks for totally-banned slurs, which automatically revoke chat privileges, while yours is looking for swear words which are acceptable in some private areas and uses different logic. Your test functions and code for checking for swear words are still separate.

I was trying to simplify things and I think I made them more confusing. Sorry about that. :-/

TheHollidayInn commented:
@abatula I believe I understand. I guess the best way to do this would be to wait for my PR to be merged, then for you to rebase on top of develop. Thanks for explaining.

abatula commented:
@TheHollidayInn I agree, that sounds like the best way to handle this.

abatula committed:
Added initial tests for blocking slurs and revoking chat priviliges

abatula committed:
Added initial tests for blocking slurs and revoking chat priviliges

abatula committed:
Uncommented line to save revoked privileges

abatula committed:
Uncommented line to save revoked privileges

abatula commented:
Does the code to email and send a slack message need to be in an api file (like the code to flag a message)? I'm trying to figure out if that's the cause of the `error: Error: Slack url is required` failure.

abatula committed:
Check that privileges are revoked in private groups

abatula committed:
Check that privileges are revoked in private groups

abatula committed:
Moved code to email/slack mods to chat api file

abatula committed:
Moved code to email/slack mods to chat api file

abatula committed:
Switched to BadRequest instead of NotFound error

abatula committed:
Switched to BadRequest instead of NotFound error

abatula committed:
Restore chat privileges after test

abatula committed:
Restore chat privileges after test

abatula commented:
I moved the code to email/slack the mods to the chat.js api file, which seems to have gotten rid of the `slack url required` error, at least.  Also switched from NotFound to BadRequest errors. 

I tried restoring chat privileges after each test so that the following tests don't fail, but that doesn't seem to be working properly yet.

abatula committed:
Using official placeholder slur

abatula committed:
Using official placeholder slur

abatula committed:
Fixed line to export sendSubscriptionNotification function for slack

abatula committed:
Fixed line to export sendSubscriptionNotification function for slack

abatula committed:
Replaced muteUser function in user methods with a single line in the chat controller file

abatula committed:
Replaced muteUser function in user methods with a single line in the chat controller file

abatula committed:
Reset chatRevoked flag to false in a single line

abatula committed:
Reset chatRevoked flag to false in a single line

abatula commented:
The current code successfully prevents users from posting chats containing slurs and removes their chat privileges (sets `flags.chatRevoked` to `true`). But when testing, the test fails on `expect(user.flags.chatRevoked).to.be.true;`, saying that `user.flags.chatRevoked` is undefined. Any idea how it could be defined (and used) in the `api.postChat` but not be defined in the test?

paglias commented on test/api/v3/integration/chat/POST-chat.test.js:
```diff
@@ -94,10 +94,10 @@ describe('POST /chat', () => {
 
-    user.flags.chatRevoked = false;
-    await user.save();
+    await user.update({'flags.chatRevoked': false});
```
consider that this only sends a command to mongo, so the database will have `flags.chatRevoked` set to false but the local representation of the user will not be updated, so `user.flags.chatRevoked` won't be `false` but still the old value

abatula committed:
Switched method of setting chatRevoked flag so that it is updated locally and in the database

abatula committed:
Switched method of setting chatRevoked flag so that it is updated locally and in the database

abatula committed:
Removed some code that got re-added after rebase

abatula committed:
Tests for automatic slur muting pass but are incomplete (do not check that chatRevoked flag is true)

abatula commented:
Tests for blocking slurs in public and private groups pass locally. Still can't directly check that the chatRevoked flag is true. It either doesn't exist, or if I set it to false at the top of the test via:
`user.flags.chatRevoked = false;`
`wait user.update({'flags.chatRevoked': false});`
it fails because it is still set to false. But if I don't use that code at the end of the test, the remaining tests will fail because chat privileges have been revoked due to the slur tests. Not sure what to make of that issue yet.

Other remaining tasks are notifications via email and slack - still in progress.

paglias commented on website/common/bannedSlurs.js:
```diff
@@ -0,0 +1,3 @@
+let bannedSlurs = ['TEST_PLACEHOLDER_SLUR_WORD_HERE'];
```
since this code is only used server side it should go in website/server not website/common, otherwise it'll be included in the client too

abatula committed:
Moved list of banned slurs to server side

abatula commented:
When the code detects a slur, it revokes chat privileges and blocks the message in all locations (public and private), but the user can continue to post messages in private groups (or at least the party chat) as long as those new messages don't get blocked for inappropriate language. Is this the desired behavior, or do we want them to be fully muted in all areas for trying to post a slur?

Alys commented:
@abatula what you've described is the desired behaviour for this issue.

At some future time we'll implement stricter levels of chat blocking and can maybe change the behaviour then. Thank you for asking! I'm sorry about the delay in our reply.

abatula committed:
Added warning to bannedSlurs file

abatula committed:
Test chat privileges revoked when posting slur in public chat

abatula committed:
Fixing rebase

abatula committed:
Fix issues left over after rebase (I hope)

abatula committed:
Added code to test for revoked chat privileges after posting a slur in a private group

abatula committed:
Moved banned slur message into locales message

abatula committed:
Added new code to check for banned slurs (parallels banned words code)

abatula commented:
Finally got the tests of public and private groups to pass. Tests that the message is blocked and chat privileges are revoked. Also made some changes to more closely mirror the recently merged code for blocking banned non-slur words.

Next on the list is sending email and slack messages to the mods.

abatula committed:
Fixed AUTHOR_MOTAL_URL in sendTxn for slur blocking

abatula committed:
Added tests that email sent on attempted slur in chat post

abatula committed:
Created context for slur-related-tests, fixed sandboxing of email. Successfully tests that email.sendTxn is called, but the email content test fails

abatula committed:
commented out slack (for now) and cleaned up tests of sending email

abatula commented:
Is there any information on how sending emails works? I'm trying to make an email with the subject "slur-report-to-mods", based on the email for flagged messages (flag-report-to-mods). But when I test, it says that the subject of the emails is "welcome" instead of what I told it to send (e.g. "slur-report-to-mods"). Even more strangely, if I tell it to use the "flag-report-to-mods", which already exists, the first test ('errors and revokes privileges when chat message contains a banned slur') says that the subject was flag-report-to-mods, but the second test ('does not allow slurs in private groups') still gets a "welcome" message.
Any pointers on where to look or what may be causing this?

Alys commented:
@abatula Is the test picking up an earlier email that's sent to the user when the user's account is created? You might need to test the second email instead of the first. (`expect(email.sendTxn.args[1][1])` instead of `expect(email.sendTxn.args[0][1])`). That's a wild guess though. :)  If nothing else, adding a (temporary?) test for a second email as well will at least let you see if there is a second email and if so, what it is. If that doesn't help, can you post here the output from the failing test?

abatula committed:
Successfully tests that slur-report-to-mods email is sent

abatula committed:
Slack message is sent, and testing works, but some user variables seem to only work when found in chat.js and passed to slack

abatula committed:
Made some fixes for lint, but not sure what to do about the camel case requirement fail, since that's how they're defined in other slack calls

abatula committed:
Slack tests pass, skipped camelcase check around those code blocks

abatula commented:
@Alys Thanks, that fixed it!

The basic functionality is now finished. Tests now check that an email is sent to the moderators when a slur is posted. It also sends and tests  the slack message. 

I have two more questions about this code:
* Some of the variables (like the user email and display name) cause errors when I try to define them in the `slack.sendSlurNotification()` function instead of passing them as arguments from the `chat.js` file. Specifically, it will cause a the following error:
 ``"code": 500
   "error": "InternalServerError"
   "message": "An unexpected error occurred."``
I'm not sure why it does this, but is the way it's set up now a problem?
* What changes would you like made to the message text/formatting of the mesages?

Alys commented on website/server/controllers/api-v3/chat.js:
```diff
@@ -141,6 +158,49 @@ api.postChat = {
+      await user.save();
+
+      // Email the mods
+      let authorEmail = user.auth.local.email;
```
Try using the `getUserInfo` function to get the email address, from `website/server/libs/email.js`. It will fetch addresses for Facebook and Google sign-in users when needed.

This might stop the InternalServerError you were getting (e.g., when there's no `local` authentication email address).

Alys commented on website/server/libs/slack.js:
```diff
@@ -93,3 +93,51 @@ module.exports = {
+
+function sendSlurNotification ({
+  // authorEmail,
+  // author,
```
This `author` line is failing because when you pass data to this function, you are using `user` as the object's name. If you change this line to `author: user,` it should work.

abatula committed:
Fixed InternalServerError caused by slack messaging

abatula committed:
Merge branch 'develop' into 8062_automatic_muting

abatula committed:
Updated chat privileges revoked error

abatula commented:
@Alys Thanks! I believe that fixed it.

I think this is no longer a work-in-progress, as the basic functionality (and corresponding tests) are complete. I'm not sure exactly what information you want contained in the emails and slack notifications, so I just took my best guess.

abatula commented:
Checking in to see what edits or changes are needed to make this PR ready.

SabreCat commented:
Ah, you're quite right, this should be switched to ready for review. I will take a look later today!

SabreCat committed:
fix(locale): typo correction

SabreCat commented:
I cleaned up a typo, but this looks *fantastic* now. @Alys, how would you like to handle adding the canonical list of banned words to website/server/bannedSlurs.js?

SabreCat commented:
Checked with @Alys offline--we'll commit adjustments to the banned word lists directly on `develop`. @abatula, thanks so much for all your work here! I've awarded you contributor tier 5 for the combination of this and recent Pixelist contributions.
